src/test/java/test/project/bookshop/config/CustomSecurityContextFactory.java

```diff
@@ -0,0 +1,29 @@
+package test.project.bookshop.config;
+
+import java.util.Set;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.test.context.support.WithSecurityContextFactory;
+import test.project.bookshop.model.Role;
+
+public class CustomSecurityContextFactory
```

Comment:
Great that you were able to create your own implementation of `@WithMockUser` annotation logic!
Can you tell why you decided to use your own implementation instead of `@WithMockUser` in your tests?